nototools/sample_with_font.py

```diff
@@ -0,0 +1,160 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Interactively generate sample text for a font.
+
+Often in bug reports people will describe text using character names.
+Replicating the sample text they describe can be a bit tedious.  This
+lets you interactively search characters in the font by name to assemble
+a string and save it to a file."""
+
+import argparse
+import codecs
+import readline
+
+from nototools import coverage
+from nototools import unicode_data
+
+from fontTools import ttLib
+
+def _build_text(name_map, initial_text=''):
+  text = initial_text
+  print 'build text using map of length %d' % len(name_map)
+  print ('enter \'quit\' to exit, '
+         '\'names\' for names, '
+         '\'dump\' to dump the current text, '
+         '\'clear\' to clear the current text, '
+         '\'write\' to be prompted for a filename to write the text to')
+  while True:
+    line = raw_input('> ')
+    if not line:
+      continue
+    if line == 'quit':
+      break
+    if line == 'names':
+      print 'names:\n  ' + '\n  '.join(sorted(name_map.keys()))
+      continue
+    if line == 'dump':
+      print 'dump: \'%s\'' % text
+      for cp in text:
+        print '%06x %s' % (ord(cp), unicode_data.name(cp))
+      continue
+    if line == 'clear':
+      text = ''
+      continue
+    if line == 'write':
+      line = raw_input('file name> ')
+      if line:
+        _write_text(line, text)
+      continue
+
+    matches = []
+    for name, cp in sorted(name_map.iteritems()):
+      if line in name:
+        matches.append(name)
+    if not matches:
+      print 'no match for "%s"'% line
+      continue
+    if len(matches) == 1:
+      print matches[0]
+      text += unichr(name_map[matches[0]])
+      continue
+
+    # if we match a full line, then use that
+    if line in matches:
+      print line
+      text += unichr(name_map[line])
+      continue
+
+    new_matches = []
+    for m in matches:
+      if line in m.split(' '):
+        new_matches.append(m)
+
+    if len(new_matches) == 1:
+      print new_matches[0]
+      text += unichr(name_map[new_matches[0]])
+      continue
+
+    if not new_matches:
+      new_matches = matches
+
+    while True:
+      print 'multiple matches:\n  ' + '\n  '.join(
+          '[%2d] %s' % (i, n) for i, n in enumerate(matches))
+      line = raw_input('0-%d or q > ' % (len(matches) - 1))
+      if line == 'q':
+        continue
+      try:
+        n = int(line)
+        if n < 0 or n >= len(matches):
+          print '%d out of range' % n
+          continue
+        text += unichr(name_map[matches[n]])
+        break
+      except:
+        continue
+
+  print 'done.'
+  return text
+
+
+def _get_char_names(charset):
+  name_map = {}
+  if charset:
+    for cp in charset:
+      name = unicode_data.name(cp)
```

Comment:
Do we want to accept/pass on PUA characters? If so, we can except a ValueError on this line.